SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex

```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
+
+While the Python debugger is useful, it does not allow us to examine implementation
+details of ITK classes on its own. A deeper investigation may be necessary for localizing
+errors in ITK classes.
+
+
+\subsubsection{SWIG C++ Runtime Tracing}
+
+As discussed in Section~\ref{subsubsec:SwigArchitecture}, SWIG wrapping generates
+C++ source files managing template implementations during ITK compilation.
+When debug symbols are available a running Python process may be attached to step through
+ITK or ITK SWIG sources at runtime. Several steps are required to set up and execute debugging:
+
+\begin{enumerate}
+  \item The ITK build must be configured so that debug symbols
+    are generated. Python wrapping must also be enabled.
+    This is accomplished by setting the CMake variables
+    \code{CMAKE\_BUILD\_TYPE:STRING="RelWithDebInfo"} and \code{ITK\_WRAP\_PYTHON:BOOL="On"}.
+    Note that the "RelWithDebInfo" build type is strongly encouraged over a "Debug"
+    build as the former will build against a standard Python distribution.
+    See Section~\ref{sec:UsingCMakeForConfiguringAndBuildingITK} for a detailed
+    explanation of how to build ITK locally with CMake.
+  \item A Python virtual environment must be appropriately configured with
+    \code{WrapITK.pth} so that the ITK debug build can be loaded.
+    See Section~\ref{sec:Wrapping} for further explanation.
+  \item ITK modules to be debugged must be loaded in a new Python session initialized
+    from the given virtual environment. Given that ITK Python uses lazy loading, it is
+    pragmatic to use \code{itk.force\_load()} to ensure that all possible debug symbols
+    are made available. The \code{os} module can be used to identify the PID of the given session.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+(venv-itk) > python
+>>> import itk
+>>> itk.force_load()
+>>> import os
+>>> os.getpid()
+99999
+\end{minted}
+
+  \item The Python session may be attached to a debugger using the returned PID.
+    \begin{itemize}
+      \item On a Windows operating system Microsoft Visual Studio 2019 or a similar platform
+        can be used for attaching to a running process for debugging. Select "Attach To Process"
+        from the Debug menu, choose "Native" code, and then search for the process PID. If
+        debug symbols were generated correctly then ITK modules will appear under the list of
+        loaded modules.
+      \item On a Linux operating system the GNU Project Debugger \code{gdb} can be used
+        for attaching to a running Python process, reading symbol files, and setting breakpoints.
+        The following example attaches to a running process and sets a breakpoint inside
+        an \code{itk.Image} Python object.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+(venv-itk) > sudo gdb
```

Comment:
No `sudo`, correct?

Reply:
The user needs certain permissions to attach to a running process, in my anecdotal experience this always seems to require superuser permissions for attaching to a running Python process.

Trying to attach without `sudo` returns an error, either when attaching with `attach` or attaching at init with `gdb -p <pid>`:
```sh
 (gdb) attach 9216
Attaching to process 9216
Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
ptrace: Operation not permitted.
```

Reply:
Interesting, was the `python` process started with the same `uid`?

Reply:
Yes, I checked back to confirm that both the `python` and `gdb` processes contain a `uid`. Looking around I can't find any other documentation stating superuser permissions as a requirement, though. I'll amend to remove `sudo` and mention that elevated permissions *may* be required.